pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/KafkaPartitionLevelConsumer.java

```diff
@@ -55,7 +58,12 @@ public MessageBatch<byte[]> fetchMessages(long startOffset, long endOffset, int
       LOGGER.debug("poll consumer: {}, startOffset: {}, endOffset:{} timeout: {}ms", _topicPartition, startOffset,
           endOffset, timeoutMillis);
     }
-    _consumer.seek(_topicPartition, startOffset);
+    Map<TopicPartition, Long> beginningOffsets = _consumer.beginningOffsets(Lists.newArrayList(_topicPartition));
+    Long beginningOffset = beginningOffsets.values().iterator().next();
+    // explicitly check for OutOfRange, where startOffset < beginningOffset
+    // without this, _consumer.poll will auto offset reset to latest, resulting in data loss
+    _consumer.seek(_topicPartition, Math.max(startOffset, beginningOffset));
```

Comment:
We're adding one more call to kafka in the execution path for all happy cases to fix a rare edge case. If Kafka consumer doesn't throw exception for out of order scenario, maybe we should check the fetched messages and in case there's no message, then we can get the beginning offset; seek to it; and then fetch again?

Reply:
agreed on the one more call to kafka for all cases. Checking the fetched messages approach won't work well either. After the auto reset and before the seek, there could be a message appearing in the topic. Plus for a consumer which sees idle time, we'd still be making the extra calls.
Can you look at this PR with the alternate approach: https://github.com/apache/pinot/pull/8309 ? This will be an opt-in approach that users who see the issue or anticipate it can turn on.

Reply:
I agree. The alternate approach in 8309 should solve the issue and it's backward compatible with no extra overhead.
👍